timeutil: Throw parsing error in case of wrong Julian date format

kakshay21 opened the pull request:
fixes: #63250
Release note (sql change): Fix Julian date parsing logic for wrong formats.

kakshay21 committed:
timeutil: Throw parsing error in case of wrong Julian date format instead of throwing 5xx

Release note (sql change): Fix Julian date parsing logic for wrong formats.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63327)
<!-- Reviewable:end -->

knz commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -765,7 +765,7 @@ func (fe *fieldExtract) Reset(field field, v int) error {
 func (fe *fieldExtract) Set(field field, v int) error {
 	if !fe.wanted.Has(field) {
-		return errors.AssertionFailedf("field %s is not wanted in %v", errors.Safe(field.Pretty()), errors.Safe(fe.wanted))
+		return inputErrorf("field %s is not wanted in %v", errors.Safe(field.Pretty()), errors.Safe(fe.wanted))
```
Can you avoid errors.Safe here

1) in the user-provided value, which is definitely unsafe

2) for the field name, instead have the field implement a SafeFormat method, or have a SafePretty method that returns redact.RedactableString

knz reviewed (commented):
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63327)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/util/timeutil/pgdate/field_extract.go, line 768 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63327#-MXmhTPgiqTA8n3nV3Ky-r1-768:-MXqnw2T8C94o7ZrbS6t:bsm6lmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b9b9d9ca1b1c9c156fc031ff5ff69166d72db4a/pkg/util/timeutil/pgdate/field_extract.go#L768)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you avoid errors.Safe here

1) in the user-provided value, which is definitely unsafe

2) for the field name, instead have the field implement a SafeFormat method, or have a SafePretty method that returns redact.RedactableString
</blockquote></details>

Ok so I have re-read the PR and I apologize for misleading you: i did not notice `errors.Safe` was already present in the code before your change.

(We need to fix this, but it is not your problem to fix)

So I think your first fix was OK. You can change it like that then I can approve the PR.


<!-- Sent from Reviewable.io -->

kakshay21 commented:
> So I think your first fix was OK. You can change it like that then I can approve the PR.
Sure no worries

kakshay21 commented:
Hi @knz not sure if `Bazel CI (Optional)` is falling because of my change

knz committed:
pgdate: sprinkle some SafeFormat methods + fix error message generation

The call to `errors.Safe(fe.wanted)` had two problems:

- we generally don't like `errors.Safe`
- it was dissimulating the `String()` method implemented
  by (`*fe.wanted`) by capturing `fe.wanted` by value,
  which caused the error message to print out `fe.wanted`
  as an integer, instead of a list of fields.

Release note (bug fix): The error messages reported when a string is
converted to time/date/timestamp and the string contains invalid
entries, now more clearly report which fields are missing or undesired.

knz committed:
pgdate: fix TestExtractInvalidJulianDate

Release note: None

knz committed:
pgdate: add more context to date parsing errors

Release note (sql change): The error payload returned to the client
when a date/time conversion fails now contains slightly more details
about the difference between the values provided and the values
that are expected.

knz reviewed (approved):
Hi Kumar,

I looked at the PR again and we are going to keep your changes. The CI failure is not your responsibility (I will look into it).

Additionally, you will notice that I have added some commits to your branch on top of your change. The goal of these commits is to generally improve error generation in that area of the code. I hope you don't mind and I hope you get a chance to learn from my additions.

Cheers

> Reviewed 1 of 2 files at r1, 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63327#-:-MY5_9Z9BE6LPTW1ftym:b-glfteg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)



<!-- Sent from Reviewable.io -->

knz commented:
(now waiting for CI success before merging)

knz commented:
@rafiss PTAL, can you review my changes on top of the contributor's

rafiss commented on pkg/util/timeutil/pgdate/field_extract_test.go:
```diff
@@ -107,6 +113,50 @@ func TestExtractSentinels(t *testing.T) {
+
+		msg := err.Error()
+		msg += "\n" + errors.FlattenDetails(err)
+		if msg != tc.expError {
```
are the error messages here stable enough be equality-checked in unit tests?

knz replied:
in this case, yes - the full text of the message is generated within this package.
(I wouldn't be comfortable including bits of messages coming from other packages)

rafiss reviewed (approved):
thank you both! i left a question, but this looks good

kakshay21 commented:
bors r+

knz commented:
bors r=rafiss

kakshay21 commented:
Thank you @knz and @rafiss
